File: backend/app/schemas/net_income.py
"""Pydantic schemas for the net-income step-function endpoints.

The model stores ``effective_month`` as an integer ``YYYYMM`` for fast,
total-orderable lookups. The HTTP boundary speaks ``"YYYY-MM"`` strings —
nicer for frontends and humans. Conversion happens here / in the router.

Amounts cross the wire as ``float`` to match the budget endpoints
(`backend/app/schemas/budget.py`). The model column itself is
``Numeric(12, 2)`` so we keep DB-side precision; we only relax to float at
the API boundary.
"""

from datetime import datetime

from pydantic import BaseModel, Field, field_validator


def _validate_year_month(value: str) -> str:
    """Validate a ``"YYYY-MM"`` string. Raises ValueError if malformed."""
    if not isinstance(value, str) or len(value) != 7 or value[4] != "-":
        raise ValueError("month must be formatted YYYY-MM")
    year_str, month_str = value.split("-")
    if not year_str.isdigit() or not month_str.isdigit():
        raise ValueError("month must be formatted YYYY-MM")
    year = int(year_str)
    month = int(month_str)
    if year < 1900 or year > 9999:
        raise ValueError("year out of range")
    if month < 1 or month > 12:
        raise ValueError("month must be 1-12")
    return value


class NetIncomePeriodResponse(BaseModel):
    """One row from the ``net_income_periods`` table."""

    id: int
    effective_month: str  # "YYYY-MM" (derived from the YYYYMM int)
    take_home_amount: float
    created_at: datetime

    model_config = {"from_attributes": False}


class NetIncomeSetRequest(BaseModel):
    """Upsert payload for ``PUT /api/net-income``."""

    effective_month: str = Field(..., description="Effective month as YYYY-MM")
    take_home_amount: float

    @field_validator("effective_month")
    @classmethod
    def _check_month(cls, v: str) -> str:
        return _validate_year_month(v)


class NetIncomeForMonthResponse(BaseModel):
    """Result of ``GET /api/net-income?month=YYYY-MM``.

    ``amount`` is the effective net income for the requested month, or
    ``null`` if no period row applies. ``from_period`` is the row that
    supplied the value (so the UI can show ``"Set in 2026-03"``), or
    ``null`` if there is no applicable row.
    """

    month: str
    amount: float | None
    from_period: NetIncomePeriodResponse | None


class PaycheckSuggestionResponse(BaseModel):
    """Result of ``GET /api/paycheck-detection/suggest``."""

    suggested_monthly_net: float | None

File: frontend/src/api/net-income.ts
/**
 * Typed fetch client for the /api/net-income and /api/paycheck-detection
 * endpoints.
 *
 * Backend contract: `backend/app/routers/net_income_router.py`. Snake_case
 * is preserved at the boundary to match `categories.ts` / `subscriptions.ts`.
 *
 * `effective_month` and `month` cross the wire as `"YYYY-MM"` strings; the
 * backend stores them as a `YYYYMM` integer internally.
 */
import { request } from "./_client";

const BASE = "/api/net-income";
const PAYCHECK_BASE = "/api/paycheck-detection";

export interface NetIncomePeriod {
  id: number;
  effective_month: string; // "YYYY-MM"
  take_home_amount: number;
  created_at: string; // ISO8601
}

export interface NetIncomeForMonth {
  month: string; // "YYYY-MM"
  amount: number | null;
  from_period: NetIncomePeriod | null;
}

export interface NetIncomeSetPayload {
  effective_month: string; // "YYYY-MM"
  take_home_amount: number;
}

export interface PaycheckSuggestion {
  suggested_monthly_net: number | null;
}

export function getNetIncome(month: string): Promise<NetIncomeForMonth> {
  const url = `${BASE}?month=${encodeURIComponent(month)}`;
  return request<NetIncomeForMonth>(url);
}

export function setNetIncome(payload: NetIncomeSetPayload): Promise<NetIncomePeriod> {
  return request<NetIncomePeriod>(BASE, {
    method: "PUT",
    body: JSON.stringify(payload)
  });
}

export function getNetIncomeHistory(): Promise<NetIncomePeriod[]> {
  return request<NetIncomePeriod[]>(`${BASE}/history`);
}

export function suggestMonthlyNet(): Promise<PaycheckSuggestion> {
  return request<PaycheckSuggestion>(`${PAYCHECK_BASE}/suggest`);
}

/** Today's month formatted as "YYYY-MM". */
export function currentMonthKey(now: Date = new Date()): string {
  const y = now.getFullYear();
  const m = String(now.getMonth() + 1).padStart(2, "0");
  return `${y}-${m}`;
}
